### ISSUE TITLE
stack overflows in traversableArray

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript-foldable-traversable#24]
For example:

```
> import Data.Array
> import Data.Traversable
> import Data.Maybe
> let xs = (1..15000)
> length xs
15000

> let ys = traverse Just xs
> isJust ys

/home/harry/documents/code/purescript-sequences/.psci_modules/node_modules/Data.Maybe/index.js:0
(function (exports, require, module, __filename, __dirname) { "use strict";
^
RangeError: Maximum call stack size exceeded
```

**Comment by garyb** [comment]
The instance has moved to `purescript-arrays` now. I think the revised version ther probably will suffer from this too though.

**Comment by hdgarrood** [comment]
Ah, ok. I guess that instance would need to move back here if we want to avoid orphans?

Do you think using Trampoline would help here? Would that be possible?

**Comment by garyb** [comment]
Damn it, you're right, it should move back. Using `Trampoline` here isn't possible due to the dependency hierarchy, but using "a trampoline" is probably the way we'll have to go.

**Comment by garyb** [comment]
Or we just do it in the FFI.

**Comment by paf31** [comment]
I don't think FFI will work for an arbitrary `Applicative`, and we have `foreachE` for `Eff` in `purescript-eff`.

**Comment by garyb** [comment]
> I don't think FFI will work for an arbitrary `Applicative`
Ah yeah, of course. I think @hdgarrood is having a look at it now anyway, or at least fighting through the jungle of prerequesites needed to be able to build the current `master`.

**Comment by hdgarrood** [comment]
Yep, more fighting than looking at this particular moment, unfortunately.

Why would FFI not work for an arbitrary applicative? This is what I've got so far: https://gist.github.com/hdgarrood/6713f157d35af74d50fe

**Comment by hdgarrood** [comment]
I've just noticed a syntax error in that, so I'm pretty sure it won't work, but is there a fundamental problem with the approach that I haven't noticed?

**Comment by paf31** [comment]
Just that I think you can build the result in constant space, but depending on the applicative, it might SO when you try to do anything with it. `Parser` maybe?

**Comment by hdgarrood** [comment]
What about `Parser`?

Would it be possible to have foldable-traversable depend on arrays, for the Foldable instance? Alternatively we could just move (or copy) foldrArray etc to here.

**Comment by garyb** [comment]
Yeah, move them back over here. They're not exported, they only exist to provide the instance.

**Comment by hdgarrood** [comment]
Fixed by #25
